File: patcha/db/retrieval/graph_context.py
"""Structural/temporal context retrieval — Layer 3, backed by the activity graph.

Answers the queries vector search can't: what happened before/after an event, what
belonged to the same work session, and what events are connected to a file/project/url.
"""

import logging
from datetime import datetime
from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    from patcha.db.activity_graph import ActivityGraph

log = logging.getLogger(__name__)


def _parse_time(value: Optional[str]) -> Optional[datetime]:
    if not value:
        return None
    try:
        return datetime.fromisoformat(value.replace("Z", "+00:00"))
    except (ValueError, AttributeError):
        log.warning("graph_context: could not parse time %r", value)
        return None


def _hhmm(ts: str) -> str:
    return (ts[:10] + " " + ts[11:16]) if len(ts) >= 16 else "??:??"


def _format_node(node: Optional[dict]) -> str:
    if not node:
        return "(unknown)"
    p = node["props"]
    ts = p.get("timestamp", "")
    etype = p.get("type", node.get("label", "?"))
    gist = p.get("gist")
    if gist:
        detail = gist
    else:
        app = p.get("app_name") or ""
        title = p.get("window_title") or ""
        detail = f"{app} — {title}" if title else (app or p.get("project") or "")
    if p.get("transition") == "switch" or p.get("trigger") in ("app_switch", "space_switch"):
        detail = f"[switch] {detail}"
    return f"[{_hhmm(ts)}] {etype}: {detail}".rstrip()


def _anchor_header(app: Optional[str], time: Optional[str]) -> str:
    parts = []
    if app:
        parts.append(f"app={app}")
    if time:
        parts.append(f"time={time}")
    return f" ({', '.join(parts)})" if parts else " (most recent)"


def get_activity_context(
    graph: "ActivityGraph",
    app: Optional[str] = None,
    time: Optional[str] = None,
    direction: str = "both",
    count: int = 3,
) -> str:
    """What was happening around an anchor event (before / after / both)."""
    anchor = graph.find_event(app=app, at=_parse_time(time))
    header = f"# Activity context{_anchor_header(app, time)}"
    if anchor is None:
        return f"{header}\nNo matching event found in the activity graph."

    anchor_id = anchor["id"]
    lines = []

    if direction in ("before", "both"):
        before = graph.before(anchor_id, count)
        for node in reversed(before):  # chronological order
            lines.append("  " + _format_node(node))

    lines.append("> " + _format_node(anchor))

    if direction in ("after", "both"):
        for node in graph.after(anchor_id, count):
            lines.append("  " + _format_node(node))

    session = graph.session_of(anchor_id)
    if session:
        apps = sorted(n["props"]["name"] for n in graph.apps_in_session(session["id"]))
        sp = session["props"]
        span = f"{_hhmm(sp.get('started_at',''))} → {_hhmm(sp.get('ended_at',''))}"
        lines.append(f"\nSession: {span} | apps: {', '.join(apps)}")

    return f"{header}\n" + "\n".join(lines)


def get_session(
    graph: "ActivityGraph",
    app: Optional[str] = None,
    time: Optional[str] = None,
) -> str:
    """Everything in the work session around an anchor (apps, events, time span)."""
    anchor = graph.find_event(app=app, at=_parse_time(time))
    header = f"# Session{_anchor_header(app, time)}"
    if anchor is None:
        return f"{header}\nNo matching event found in the activity graph."

    session = graph.session_of(anchor["id"])
    if session is None:
        return f"{header}\nAnchor event has no session."

    sp = session["props"]
    span = f"{_hhmm(sp.get('started_at',''))} → {_hhmm(sp.get('ended_at',''))}"
    apps = sorted(n["props"]["name"] for n in graph.apps_in_session(session["id"]))
    events = sorted(
        graph.events_in_session(session["id"]),
        key=lambda n: n["props"].get("ts_ms", 0),
    )
    lines = [f"Span: {span}", f"Apps: {', '.join(apps)}", f"Events ({len(events)}):"]
    lines += ["  " + _format_node(n) for n in events]
    return f"{header}\n" + "\n".join(lines)


def find_connected(
    graph: "ActivityGraph",
    file: Optional[str] = None,
    project: Optional[str] = None,
    url: Optional[str] = None,
    app: Optional[str] = None,
) -> str:
    """Events structurally connected to a file, project, url, or app (multi-hop)."""
    if file:
        node_id, label = f"file::{file}", f"file '{file}'"
    elif project:
        node_id, label = f"project::{project}", f"project '{project}'"
    elif url:
        node_id, label = f"url::{url}", f"url '{url}'"
    elif app:
        node_id, label = f"app::{app}", f"app '{app}'"
    else:
        return "# Connected activity\nProvide one of: file, project, url, app."

    header = f"# Connected activity — {label}"
    if graph.get_node(node_id) is None:
        return f"{header}\nNo such node in the activity graph."

    events = sorted(
        graph.events_touching(node_id),
        key=lambda n: n["props"].get("ts_ms", 0),
    )
    if not events:
        return f"{header}\nNode exists but no events reference it."

    lines = [f"Events ({len(events)}):"] + ["  " + _format_node(n) for n in events]
    return f"{header}\n" + "\n".join(lines)

File: rust/patcha/src/db/retrieval/graph_context.rs
use crate::{
    db::activity_graph::ActivityGraph,
    models::GraphNode,
};
use anyhow::Result;
use chrono::{DateTime, Utc};

// ---------------------------------------------------------------------------
// Public API (mirrors Python patcha/db/retrieval/graph_context.py)
// ---------------------------------------------------------------------------

pub fn get_activity_context(
    graph: &ActivityGraph,
    app: Option<&str>,
    time: Option<&str>,
    direction: &str,
    count: usize,
) -> Result<String> {
    let at = parse_time(time);
    let header = format!("# Activity context{}", anchor_header(app, time));
    let anchor = graph.find_event(app, at)?;
    let Some(anchor) = anchor else {
        return Ok(format!("{header}\nNo matching event found in the activity graph."));
    };

    let mut lines: Vec<String> = Vec::new();

    if direction == "before" || direction == "both" {
        let before = graph.events_before(&anchor.id, count)?;
        for node in before.into_iter().rev() {
            lines.push(format!("  {}", format_node(&node)));
        }
    }

    lines.push(format!("> {}", format_node(&anchor)));

    if direction == "after" || direction == "both" {
        let after = graph.events_after(&anchor.id, count)?;
        for node in after {
            lines.push(format!("  {}", format_node(&node)));
        }
    }

    if let Ok(Some(session_id)) = graph.session_of(&anchor.id) {
        let session_node = graph.get_node(&session_id)?;
        let apps = graph.apps_in_session(&session_id).unwrap_or_default();
        let mut sorted_apps = apps;
        sorted_apps.sort();
        let span = session_node
            .as_ref()
            .map(|n| {
                let start = n
                    .props
                    .get("start_ts")
                    .and_then(|v| v.as_str())
                    .unwrap_or("");
                hhmm(start)
            })
            .unwrap_or_else(|| "??:??".into());
        lines.push(format!(
            "\nSession: {span} | apps: {}",
            sorted_apps.join(", ")
        ));
    }

    Ok(format!("{header}\n{}", lines.join("\n")))
}

pub fn get_session(
    graph: &ActivityGraph,
    app: Option<&str>,
    time: Option<&str>,
) -> Result<String> {
    let at = parse_time(time);
    let header = format!("# Session{}", anchor_header(app, time));
    let anchor = graph.find_event(app, at)?;
    let Some(anchor) = anchor else {
        return Ok(format!("{header}\nNo matching event found in the activity graph."));
    };

    let session_id = match graph.session_of(&anchor.id)? {
        Some(id) => id,
        None => return Ok(format!("{header}\nAnchor event has no session.")),
    };

    let session_node = graph.get_node(&session_id)?;
    let span = session_node
        .as_ref()
        .map(|n| {
            let start = n
                .props
                .get("start_ts")
                .and_then(|v| v.as_str())
                .unwrap_or("");
            hhmm(start)
        })
        .unwrap_or_else(|| "??:??".into());

    let mut apps = graph.apps_in_session(&session_id).unwrap_or_default();
    apps.sort();

    let event_ids = graph.events_in_session(&session_id).unwrap_or_default();
    let mut events: Vec<GraphNode> = event_ids
        .iter()
        .filter_map(|id| graph.get_node(id).ok().flatten())
        .collect();
    events.sort_by_key(|n| {
        n.props
            .get("ts_ms")
            .and_then(|v| v.as_i64())
            .unwrap_or(0)
    });

    let mut lines = vec![
        format!("Span: {span}"),
        format!("Apps: {}", apps.join(", ")),
        format!("Events ({}):", events.len()),
    ];
    for node in &events {
        lines.push(format!("  {}", format_node(node)));
    }

    Ok(format!("{header}\n{}", lines.join("\n")))
}

pub fn find_connected(
    graph: &ActivityGraph,
    file: Option<&str>,
    project: Option<&str>,
    url: Option<&str>,
    app: Option<&str>,
) -> Result<String> {
    let (target, target_type, label) = if let Some(f) = file {
        (f, "File", format!("file '{f}'"))
    } else if let Some(p) = project {
        (p, "Project", format!("project '{p}'"))
    } else if let Some(u) = url {
        (u, "URL", format!("url '{u}'"))
    } else if let Some(a) = app {
        (a, "App", format!("app '{a}'"))
    } else {
        return Ok("# Connected activity\nProvide one of: file, project, url, app.".into());
    };

    let header = format!("# Connected activity — {label}");
    let mut events = graph.events_touching(target, target_type)?;
    if events.is_empty() {
        return Ok(format!("{header}\nNo events connected to this {target_type}."));
    }

    events.sort_by_key(|n| {
        n.props
            .get("ts_ms")
            .and_then(|v| v.as_i64())
            .unwrap_or(0)
    });

    let mut lines = vec![format!("Events ({}):", events.len())];
    for node in &events {
        lines.push(format!("  {}", format_node(node)));
    }

    Ok(format!("{header}\n{}", lines.join("\n")))
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn parse_time(value: Option<&str>) -> Option<DateTime<Utc>> {
    value.and_then(|s| {
        DateTime::parse_from_rfc3339(&s.replace('Z', "+00:00"))
            .ok()
            .map(|dt| dt.with_timezone(&Utc))
    })
}

fn hhmm(ts: &str) -> String {
    if ts.len() >= 16 {
        format!("{} {}", &ts[..10], &ts[11..16])
    } else {
        "??:??".into()
    }
}

fn format_node(node: &GraphNode) -> String {
    let p = &node.props;
    let ts = p.get("timestamp").and_then(|v| v.as_str()).unwrap_or("");
    let etype = p
        .get("type")
        .and_then(|v| v.as_str())
        .unwrap_or(node.label.as_str());

    let detail = if let Some(gist) = p.get("gist").and_then(|v| v.as_str()) {
        gist.to_owned()
    } else {
        let app = p.get("app_name").and_then(|v| v.as_str()).unwrap_or("");
        let title = p
            .get("window_title")
            .and_then(|v| v.as_str())
            .unwrap_or("");
        let proj = p.get("project").and_then(|v| v.as_str()).unwrap_or("");
        if !title.is_empty() {
            format!("{app} — {title}")
        } else if !app.is_empty() {
            app.to_owned()
        } else if !proj.is_empty() {
            proj.to_owned()
        } else {
            String::new()
        }
    };

    format!("[{}] {}: {}", hhmm(ts), etype, detail).trim_end().to_owned()
}

fn anchor_header(app: Option<&str>, time: Option<&str>) -> String {
    let mut parts = Vec::new();
    if let Some(a) = app {
        parts.push(format!("app={a}"));
    }
    if let Some(t) = time {
        parts.push(format!("time={t}"));
    }
    if parts.is_empty() {
        " (most recent)".into()
    } else {
        format!(" ({})", parts.join(", "))
    }
}
